### PR TITLE
remove getWeights from Function interface.

### PR DESCRIPTION
**Pull request by jq**
@deerzq @yolken @hectorgon @spencerde @christhetree

### CODE REVIEW

File: core/src/main/java/com/airbnb/aerosolve/core/function/Spline.java [hunk 1]
```diff
@@ -21,21 +21,23 @@ public Spline(float minVal, float maxVal, float [] weights) {
...
-        float x = diff * t + other.minVal;
-        newWeights[i] = other.evaluate(x);
-      }
+      setupSpline(other.minVal, other.maxVal, other.newBinWeights(newBins));
+    }
+  }
+
+  public float[] newBinWeights(int newBins) {
+    float[] newWeights = new float[newBins];
+    float scale = 1.0f / (newBins - 1.0f);
```
Comment (spencerde) [inline core/src/main/java/com/airbnb/aerosolve/core/function/Spline.java]: `newBins` won't ever be 1 here?

File: core/src/main/java/com/airbnb/aerosolve/core/function/Spline.java [hunk 2]
```diff
@@ -21,21 +21,23 @@ public Spline(float minVal, float maxVal, float [] weights) {
...
-      for (int i = 0; i < newBins; i++) {
-        float t = i * scale;
-        float x = diff * t + other.minVal;
-        newWeights[i] = other.evaluate(x);
-      }
+      setupSpline(other.minVal, other.maxVal, other.newBinWeights(newBins));
+    }
+  }
+
+  public float[] newBinWeights(int newBins) {
```
Comment (spencerde) [inline core/src/main/java/com/airbnb/aerosolve/core/function/Spline.java]: Can you add a comment for what this does, that it generates a new set of weights which mimics the shape of this spline rescaled to a new number of bins?

**Comment by spencerde** [comment]
One style comment, overall LGTM

**Comment by jq** [comment]
@spencerde thanks for review, add comments and name it to weightsByNumBins

File: core/src/main/java/com/airbnb/aerosolve/core/function/Linear.java [hunk 3]
```diff
@@ -28,6 +26,20 @@ public Linear(float minVal, float maxVal, float[] weights) {
   }
 
   @Override
+  public Function aggregate(Iterable<Function> functions, float scale, int numBins) {
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/function/Linear.java]: numBins is not used here.

Comment (jq) [inline core/src/main/java/com/airbnb/aerosolve/core/function/Linear.java]: yes, just to keep interface for both linear and spline

**Comment by deerzq** [comment]
We may want to rename numBins to something else, since it's a Spline specific thing. LGTM

**Pull request merged**